Add a function to request a recommended max number of Domains

haesbaert opened the pull request:
This adds `Domain.recommended_domains ()` (better name to be discussed) which returns a reasonable max number of parallel units of execution, this number should express the warm value of what the system could possibly run in parallel "at that time", so we're interested in how many logical CPUs the system can give us, not necessarily how many CPUs the system has.

The question "How many Domains should I spawn ?" seems to come often in the ocaml-multicore discussions, this attempts to address it without being too invasive.

There is no portable way to retrieve such information, so there are 4 methods

1. GNU `pthread_getaffinity_np(3)` (Linux)
2. BSD `pthread_getaffinity_np(3)` (FreeBSD)
3. `sysconf(_SC_NPROCESSORS_ONLN)` (Darwin, NetBSD, OpenBSD & others)
4. `sysinfo.dwNumberOfProcessors` (Windows)

`pthread_getaffinity_np(3)` is listed twice since the API is slightly different for Linux and FreeBSD, NetBSD doesn't have all the FreeBSD macros so it will just fallback onto `sysconf`, OpenBSD doesn't have it.

`pthread_getaffinity_np(3)` can return a smaller set than `sysconf(_SC_NPROCESSORS_ONLN)`, the child process inherits the affinity of the parent, on `linux` this can be observed with tasket:
```
rec.exe                                                                                                    
recommended_domains 8

taskset 0xf rec.exe                                                                                        
recommended_domains 4
```

I had trouble coming up with a good name, other ideas: `parallel_units` `max_parallel_units` `max_exec_units` `domain_num_hint` `max_domain_hint`. If this gets any traction I'll adjust the commit message accordingly.

haesbaert committed:
Domain.recommended_domains

Or whatever name we come up with.

nojb commented:
xref discussion starting at https://github.com/ocaml/ocaml/issues/10971#issuecomment-1025751321

dbuenzli commented:
This is not only useful for domains, can't we rather have this as `Sys.processor_count : unit -> int` ?

haesbaert commented:
> This is not only useful for domains, can't we rather have this as `Sys.processor_count : unit -> int` ?

We could, actually that was my first approach, but after talking to some people it seemed the general preference was to have it within Domain. If we choose to go the Sys route it might make sense to have a `Sys.processor_count` and `Sys.processor_online_count`. I'm happy with either.

dra27 commented:
The intention here was not to be committed to providing an accurate metric for (logical) processor count, but just to be limited to dealing with a recommended number of domains which should be spawned (especially trying to take limitations and so forth into account).

dra27 commented:
Assuming there's no particular cost to it, I was guessing it would be OK for this just to be `int` and computed at startup (along with various other sysconf, etc. probes which already have to take place)

nojb commented:
> The intention here was not to be committed to providing an accurate metric for (logical) processor count, but just to be limited to dealing with a recommended number of domains which should be spawned (especially trying to take limitations and so forth into account).

I think this is being too careful. As far as I know, everywhere where this function is exposed it is defined more or less the same way as in this PR. I would just expose it as `Sys.processor_count`, and if some more Domains-specific logic is needed a wrapper around it can be added and exposed as `Domains.recommended_domains`...

dbuenzli commented:
> The intention here was not to be committed to providing an accurate metric for (logical) processor count

Why not ? Personally I'd rather have that.

I had to implement this function more than once (and in a likely cross-platform broken way). It's good if the system can provide it. Besides as can be seen in the linked discussion other languages out there do actually provide you with that information.

haesbaert commented:
> Assuming there's no particular cost to it, I was guessing it would be OK for this just to be `int` and computed at startup (along with various other sysconf, etc. probes which already have to take place)

The cost is basically CPU_COUNT() which has to count bits, the default set on Linux is 1024 bits, the other point is that it would go a bit against giving a warm/fresh view of it.

haesbaert commented:
> The intention here was not to be committed to providing an accurate metric for (logical) processor count, but just to be limited to dealing with a recommended number of domains which should be spawned (especially trying to take limitations and so forth into account).

I'd point out that giving an accurate total logical processor count is easier (as in always accurate and more portable) than giving a "current" logical processor count. The other argument is that if you're doing multi-process things, it's still sensible to be able to query logical count (which looks a little odd if you extrapolate that from Domain).

dra27 commented:
Fundamentally, I don't mind if we end up with a function in `Sys`, but if we're going to report the actual number of logical processors, then it does need to be correct. For example, on my Windows system here, `Domain.recommended_domains ()` is returning `64` which is the correct maximum number of domains to spin up, but _not_ the correct number of online or available logical processors (the code to get the correct `128` is non-trivial).

If we do put a function in `Sys`:

> if some more Domains-specific logic is needed a wrapper around it can be added and exposed as `Domains.recommended_domains`...

This will be necessary, partly for the Windows example above but more relevantly because `caml_recommended_domains` already ought to be returning the minimum of the available cpus and `Max_domains` (which is only 16 on a 32bit system).

haesbaert committed:
Curb recommended_domains to Max_domains

haesbaert commented:
Since there is a `Max_domains` I'd argue that if we want `Sys.processor_count` we can bring it later, and then define `recommended_domains` in terms of `processor_count`. We do need something that respects `Max_domain` and also takes number of CPUs into account sooner.

I think we might end up providing both in the end.

nojb commented:
> the code to get the correct `128` is non-trivial.

Good catch! I extracted the following code from MSDN (no error checking, and don't have a 128 core machine to test :))

https://gist.github.com/nojb/8306e10b6fef54b6b4baf16b4218ac2f

haesbaert commented:
Python does this for windows: `GetActiveProcessorCount(ALL_PROCESSOR_GROUPS)`
https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getactiveprocessorcount

dra27 commented:
@nojb - that's still processor group affected, so gives 64. It can be got with `GetLogcalProcessorInformationEx` which can give processor group information, but it's moot for the recommended number of domains which does need to be 64, without more scheduling work (which I hope/assume Python for Windows does, @haesbaert, or is that just for a sysinfo function?!). The point is that the _present_ implementation of `Domain.recommended_domains` gives a number which is not the logical cpu count. One of us will hopefully sort out the thread scheduling in the future to allow more than 64 domains on Windows, at which point that _future_ version of `Domain.recommended_domains` could then start returning the actual logical cpu count, even on Windows.

dra27 commented:
(in all of this, I'm in no way against us having a function which returns the logical CPU count, my only point is that the logical cpu count is not portably related to the maximum number of domains you ought to try to spin up on a given machine, and having a meaningful recommendation for that number does seem important)

haesbaert commented:
@dra27 Python would be returning 128 for `os_cpu_count`, which is like a sysinfo, they even clarify it:
```
os.cpu_count

Return the number of CPUs in the system; return None if indeterminable.

This number is **not** equivalent to the number of CPUs the current process can
use.  The number of usable CPUs can be obtained with
``len(os.sched_getaffinity(0))``
```

They don't implement a `os.sched_getaffinitity()` for windows and they just throw a Not_implemented.

haesbaert committed:
Domain.recommended_domains () -> Domain.recommended_domains as suggested

kayceesrk commented on stdlib/domain.ml:
```diff
@@ -25,6 +25,8 @@ module Raw = struct
     = "caml_ml_domain_id"
   external cpu_relax : unit -> unit
     = "caml_ml_domain_cpu_relax"
+  external get_recommended_domains: unit -> int
```
This function may be marked as [@@noalloc]
```suggestion
  external get_recommended_domains: unit -> int [@@noalloc]
```

kayceesrk replied:
Not specific to this PR, but `cpu_relax` and `self` should also be made [@@noalloc].

gasche commented:
Silly naming input: to me a variable called `<adjective>_domains` suggests that it is going to return domains (maybe a `Domain.t list`? A name such as `recommended_domain_count` (or `_number`, etc.) would be more natural to me.

haesbaert committed:
Mark recommended_domains as [@@noalloc]

As suggested by KC

haesbaert commented:
> Silly naming input: to me a variable called `<adjective>_domains` suggests that it is going to return domains (maybe a `Domain.t list`? A name such as `recommended_domain_count` (or `_number`, etc.) would be more natural to me.

I agree, I do think it's a bit long though. `domain_count_hint` comes to mind as well

dbuenzli commented:
You could also try to drop the `domain` since it's already in the prefix. For example:

```
Domain.max_spawn_hint
Domain.max_count_hint
```

haesbaert commented:
> You could also try to drop the `domain` since it's already in the prefix. For example:
> 
> ```
> Domain.max_spawn_hint
> Domain.max_count_hint
> ```

I like `max_spawn_hint`, it will also more often than not be used closely to `Domain.spawn ()` calls and it's not ambiguous like `recommended_domains`. I'd vote for `max_spawn_hint` sofar.

gasche commented:
Personally I don't mind the long name (why would we have more than one occurrence in a program anyway?) and I find `{Domain,Sys}.recommended_domain_count` clearer than `Domain.max_spawn_hint`.

haesbaert commented:
Is the name now the major blocker for this ?
I'd be happy with either `Domain.max_spawn_hint` or `{Domain,Sys}.recommended_domain_count`.

dbuenzli commented:
I don't think one wants to have `Sys.recommend_domain_count`. I thought the functions were to be separated in two, one for the system processor count, e.g. `Sys.processor_count` that reports a property of your hardware/system (to be also used e.g. for programs that do multiprocessing) and one for maximal domain count which could be an actual function of that value.

haesbaert commented:
> I don't think one wants to have `Sys.recommend_domain_count`. I thought the functions were to be separated in two, one for the system processor count, e.g. `Sys.processor_count` that reports a property of your hardware/system (to be also used e.g. for programs that do multiprocessing) and one for maximal domain count which could be an actual function of that value.

Yes, scratch the `Sys.recommend_domain_count`, I can work on a `Sys.processor_count` (I have the code already) or equivalent once we decide how to proceed with `Domain.recommended_domain_count` or `Domain.max_spawn_hint`.

nojb commented:
> > I don't think one wants to have `Sys.recommend_domain_count`. I thought the functions were to be separated in two, one for the system processor count, e.g. `Sys.processor_count` that reports a property of your hardware/system (to be also used e.g. for programs that do multiprocessing) and one for maximal domain count which could be an actual function of that value.
> 
> Yes, scratch the `Sys.recommend_domain_count`, I can work on a `Sys.processor_count` (I have the code already) or equivalent once we decide how to proceed with `Domain.recommended_domain_count` or `Domain.max_spawn_hint`.

Let's go with `recommended_domain_count`, there's no need to be cryptic and `max_spawn_hint` does not make much grammatical sense.

haesbaert committed:
Rename recommended_domains to recommended_domain_count

haesbaert committed:
Add a Domain.recommended_domain_count entry to Changes

I hope it's alright that I've added the reviewers already.

haesbaert committed:
Add tests for Domain.recommended_count

I've added the stub just to fetch Max_domains which is not exposed, I also
defined CAML_INTERNAL, I hope this is ok on tests.

The test is very defensive, we try external tools that should return a number
that is atleast larger than the one we computed, instead of doing a bunch of OS
dependent stuff, just try them all:

 * nproc should work on linux
 * sysctl hw.ncpu should work on all BSDs including Darwin, hw.ncpu is the old
school "big number", so it's the largest possible value (as in doesn't take into
account affinity or online cpus).

In case no tools are available (no nproc, no sysctl hw.ncpu) assume the test
passed (still check max and min though), we can't infer the environment we're
being built, it might be chrooted and etcetera.

The split into two tests is so that we can at least test the very basic stuff on
windows.

haesbaert commented:
I've added the entry to Changes and wrote a regression test, besides I've tested myself on linux/OS X/FreeBSD/Windows.
Best I could think test-wise is relying on an external tool like `nproc` or `sysctl hw.ncpu` and making sure `recommended_comain_count` returns less of equal that. The test is split in two bits since one depends on `Unix`.

nojb commented on runtime/domain.c:
```diff
@@ -18,10 +18,23 @@
 
 #define CAML_INTERNALS
 
+#define _GNU_SOURCE  /* For sched.h CPU_ZERO(3) and CPU_COUNT(3) */
```
Is this only needed when `HAS_GNU_GETAFFINITY_NP` is defined?

haesbaert replied:
> Is this only needed when `HAS_GNU_GETAFFINITY_NP` is defined?

Yes and no, the problem is that in order to get `HAS_GNU_GETAFFINITY_NP`, I need to include `"caml/config.h"`, but I have to define `_GNU_SOURCE` before including `caml/config.h` since it pulls other headers that break `_GNU_SOURCE` since it must be included before things like `stdint.h` if you check `runtime/{domain.c,floats.c,unix.c}` they do the same.

nojb replied:
Got it, thanks!

nojb reviewed (approved):
LGTM

I only wonder about the `recommended_domain_count.ml` test: personally it feels a bit overkill, I would drop it, but up to you :)

nojb commented:
Given that this touches the stdlib, we need a scond approval. Perhaps @dra27 can do the honours?

haesbaert commented:
> LGTM
> 
> I only wonder about the `recommended_domain_count.ml` test: personally it feels a bit overkill, I would drop it, but up to you :)

Great \o/, I was on the fence for this as well, I'd say keep it and I'll keep an eye if anything fishy shows up (I don't think it can hurt)

dra27 commented on runtime/domain.c:
```diff
@@ -1754,3 +1767,37 @@ CAMLprim value caml_domain_dls_get(value unused)
+
+CAMLprim value caml_recommended_domain_count(value unused)
+{
+  CAMLnoalloc;
```
This isn't needed for this function:
```suggestion
```

dra27 commented on runtime/domain.c:
```diff
@@ -1754,3 +1767,37 @@ CAMLprim value caml_domain_dls_get(value unused)
+CAMLprim value caml_recommended_domain_count(value unused)
+{
+  CAMLnoalloc;
+  long n = -1;
```
Not that we'll see a machine where it matters, but 🙂
```suggestion
  intnat n = -1;
```

dbuenzli commented:
So we don't get the function in `Sys` in the end ?

dra27 reviewed (commented):
Two nits, but otherwise fine for me! It's running through [precheck#732](https://ci.inria.fr/ocaml/job/precheck/732).

FWIW I'd put the primitive in `runtime/sys.c` and the test in `testsuite/tests/lib-domains`, but neither is critical.

nojb commented:
> So we don't get the function in `Sys` in the end ?

If I understood well, there was consensus on adding the function to `Sys` as well, but that was a bit more work, so the idea is to unlock this PR now, and add the function in `Sys` in a later PR.

haesbaert commented:
> So we don't get the function in `Sys` in the end ?

I can work on it in another PR, I think it makes more sense to get this in first so we don't block this one

haesbaert committed:
Update runtime/domain.c

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

haesbaert committed:
Update runtime/domain.c

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

dra27 reviewed (approved):
Precheck seems OK (or at least the errors are clearly unrelated to this PR).

Good to go, thanks!

gasche commented:
I realize only now that this primitive is currently exposed as

```ocaml
val recommended_domain_count : int
(** The recommended maximum number of domains to be spawned. *)
```

I think that making this a constant rather than a function is very risky. It means that we give up forever on making a part of the decision at runtime. I propose to change this back to `unit -> int`. What do you think?

xavierleroy commented:
>  It means that we give up forever on making a part of the decision at runtime

The decision is taken at program start-up time, but yes, I see what you mean and I'm in favor of `unit -> int`.

haesbaert commented:
I agree, this has been on the back of my mind for a while.

On Wed, Sep 14, 2022, 08:32 Xavier Leroy ***@***.***> wrote:

> It means that we give up forever on making a part of the decision at
> runtime
>
> The decision is taken at program start-up time, but yes, I see what you
> mean and I'm in favor of unit -> int.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/ocaml/pull/11309#issuecomment-1246305154>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABR2EBQUI2UELDAAHFCL4LV6FWRXANCNFSM5YKRZ67A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dbuenzli commented:
> > So we don't get the function in `Sys` in the end ?
> 
> I can work on it in another PR, I think it makes more sense to get this in first so we don't block this one

@haesbaert :–)

haesbaert commented:
> > > So we don't get the function in `Sys` in the end ?
> > 
> > 
> > I can work on it in another PR, I think it makes more sense to get this in first so we don't block this one
> 
> @haesbaert :–)

Apologies, I haven't forgotten :D just got otherwise busy.
